winton_kafka_streams/processor/_stream_thread.py

```diff
@@ -93,20 +93,30 @@ def __init__(self, _topology, _config, _kafka_supplier):
         self.consumer = self.kafka_supplier.consumer()
 
         self.thread = threading.Thread(target=self.run)
+        self.state_listener = None
         self.set_state(self.State.RUNNING)
 
+    def thread_id(self):
+        return self.thread.ident
+
     def set_state(self, new_state):
         old_state = self.state
         if not old_state.valid_transition_to(new_state):
             log.warn(f'Unexpected state transition from {old_state} to {new_state}.')
         else:
             log.info(f'State transition from {old_state} to {new_state}.')
         self.state = new_state
+        if self.state_listener:
+            self.state_listener(self, old_state, new_state)
 
     def set_state_when_not_in_pending_shutdown(self, new_state):
         if not self.state is self.State.PENDING_SHUTDOWN:
             self.set_state(new_state)
 
+    def set_state_listener(self, listener):
```

Comment:
Just use the variable directly instead of going via set function (looking at too much Java?!). Alternatively if you want to indicate the listener is for internal use only then name the variable as self._state_listener?